File: Subprogramas como parâmetros/subprogramas-como-parametros.php
<?php
# Em PHP, uma variável contendo uma string pode ser usada para ativar
# um subprograma. Para passar um subprograma como parâmetro, basta passar
# uma string com o nome do subprograma.

# Operações de comparação entre funcionários.
function MenorNome($func1, $func2) {
    return $func1['nome'] < $func2['nome'];
}
function MenorSalario($func1, $func2) {
    return $func1['salario'] < $func2['salario'];
}

function Ordenar(&$vetor, $criterio) {
    # Insertionsort com critério de ordenação.
    # $criterio é uma função que retorna verdadeiro se o primeiro parâmetro
    #     é menor que o segundo.
    $tamanho = count($vetor);
    for ($i = 1; $i < $tamanho; ++$i) {
        $aux = $vetor[$i]; // numero que sera inserido na parte ordenada
        $j = $i-1;
        while ($j >= 0 and $criterio($aux, $vetor[$j])) {
            $vetor[$j+1] = $vetor[$j];
            --$j;
        }
        $vetor[$j+1] = $aux;
    }
}

function EscreverFuncionario($func) {
    echo "(${func['nome']}, ${func['salario']})";
}

function EscreverFuncionarios($vetor) {
    echo "[";
    $ultimo = count($vetor)-1;
    for ($i = 0; $i < $ultimo; ++$i) {
        EscreverFuncionario($vetor[$i]);
        echo ", ";
    }
    if ($ultimo >= 0)
        EscreverFuncionario($vetor[$ultimo]);
    echo "]";
}

# Criar 3 funcionários
# Em PHP não precisamos de registros porque temos vetores associativos.
$funcionario1 = array('nome' => 'Joao da Silva', 'salario' => 2012.0);
$funcionario2 = array('nome' => 'Jose Santos', 'salario' => 1930.0);
$funcionario3 = array('nome' => 'Carlos Oliveira', 'salario' => 2420.0);

# Criar um vetor de funcionários
$funcionarios = array($funcionario1, $funcionario2, $funcionario3);

Ordenar($funcionarios, 'MenorSalario');
echo "Ordenado por salario:\n";
EscreverFuncionarios($funcionarios);
Ordenar($funcionarios, 'MenorNome');
echo "\nOrdenado por nome:\n";
EscreverFuncionarios($funcionarios);
?>

File: Subprogramas como parâmetros/subprogramas-como-parametros.cpp
// Este programa é um exemplo de passagem de subprogramas como parâmetros. O recurso
// usado aqui é o uso de ponteiro para subprograma. Existem duas outras formas de se
// fazer isso em C++, com maior legibilidade, porém essas outras formas dependem de
// bibliotecas ou orientação a objetos.

// Este programa também contém exemplo de sobrecarga de operador e subprograma genérico.

#include <iostream>
#include <vector>
#include <string>

using namespace std;

struct Funcionario {
    string nome;
    float salario;
};

// Sobrecarga do operador de escrita para escrever funcionarios.
ostream& operator<< (ostream& saida, const Funcionario& func) {
    return saida << "[" << func.nome << ", " << func.salario << "]";
}

// Comparação de funcionarios pelo nome
bool MenorNome(const Funcionario& f1, const Funcionario& f2) {
    return f1.nome < f2.nome;
}

// Comparação de funcionarios pelo salario
bool MenorSalario(const Funcionario& f1, const Funcionario& f2) {
    return f1.salario < f2.salario;
}

// Ordenação de vetor de qualquer coisa (subprograma genérico) que recebe um critério
// de ordenação que é uma função. O parâmetro "Menor" é uma função que recebe dois
// parâmetros do tipo "Tipo" por referência constante e retorna um bool.
template <typename Tipo>
void Ordenar(vector<Tipo>& vetor, bool (*Menor)(const Tipo&, const Tipo&)) {
    unsigned tamanho = unsigned(vetor.size());
    for (unsigned i = 1; i < tamanho; ++i) {
        Tipo aux = vetor[i]; // numero que sera inserido na parte ordenada
        int j = i-1;
        while (j >= 0 and Menor(aux, vetor[j])) {
            vetor[j+1] = vetor[j];
            --j;
        }
        vetor[j+1] = aux;
    }
}

int main() {
    vector<Funcionario> funcionarios = {
        {"Joao da Silva", 2012}, {"Jose Santos", 1930}, {"Carlos Oliveira", 2420}};

    Ordenar(funcionarios, MenorSalario);
    cout << "Funcionarios ordenados por salario:\n";
    for (Funcionario& func: funcionarios)
        cout << func << ' ';
    cout << endl;

    Ordenar(funcionarios, MenorNome);
    cout << "Funcionarios ordenados por nome:\n";
    for (Funcionario& func: funcionarios)
        cout << func << ' ';
    cout << endl;

    return 0;
}

